src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,283 @@
+                _structLayoutAttribute = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesStructLayoutAttribute);
+                _lcidConversionAttribute = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesLCIDConversionAttribute);
+                if (compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesMarshal, out var marshalType))
+                {
+                    var marshalMethods = ImmutableArray.CreateBuilder<ISymbol>();
+                    marshalMethods.AddRange(marshalType.GetMembers("SizeOf"));
+                    marshalMethods.AddRange(marshalType.GetMembers("OffsetOf"));
+                    marshalMethods.AddRange(marshalType.GetMembers("StructureToPtr"));
+                    marshalMethods.AddRange(marshalType.GetMembers("PtrToStructure"));
+                    _marshalMethods = marshalMethods.ToImmutable();
+                }
+            }
+
+            public void RegisterActions(CompilationStartAnalysisContext context)
+            {
+                context.RegisterSymbolAction(AnalyzeMethod, SymbolKind.Method);
+
+                context.RegisterOperationAction(AnalyzeMethodCall, OperationKind.Invocation);
+
+                context.RegisterOperationAction(AnalyzeFunctionPointerCall, OperationKindEx.FunctionPointerInvocation);
+
+                if (_unmanagedFunctionPointerAttribute is not null)
+                {
+                    context.RegisterSymbolAction(AnalyzeType, SymbolKind.NamedType);
+                }
+            }
+
+            public void AnalyzeMethodCall(OperationAnalysisContext context)
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                if (_marshalMethods.Contains(invocation.TargetMethod.ConstructedFrom))
+                {
+                    bool canTransformToDisabledMarshallingEquivalent = CanTransformToDisabledMarshallingEquivalent(invocation);
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(
+                        MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabled,
+                        ImmutableDictionary.Create<string, string?>().Add(CanConvertToDisabledMarshallingEquivalentKey, canTransformToDisabledMarshallingEquivalent ? "true" : null),
+                        invocation.TargetMethod.ToDisplayString()));
+                }
+            }
+
+            private static bool CanTransformToDisabledMarshallingEquivalent(IInvocationOperation invocation)
+            {
+                return invocation.TargetMethod.Name switch
+                {
+                    "OffsetOf" => false,
+                    "SizeOf" => invocation.TargetMethod.IsGenericMethod || invocation.Arguments[0].Value is ITypeOfOperation { TypeOperand.IsUnmanagedType: true },
```

Comment:
I don't think accessing the first element is safe. The invocation may not have a parameter (error scenario).